src/openms/source/METADATA/ID/IdentificationData.cpp

```diff
@@ -93,23 +93,41 @@ namespace OpenMS
 
 
   IdentificationData::InputFileRef
-  IdentificationData::registerInputFile(const String& file)
+  IdentificationData::registerInputFile(const InputFile& file)
   {
-    return input_files_.insert(file).first;
+    if (!no_checks_ && file.name.empty()) // key may not be empty
+    {
+      String msg = "input file must have a name";
+      throw Exception::IllegalArgument(__FILE__, __LINE__,
+                                       OPENMS_PRETTY_FUNCTION, msg);
+    }
+    auto result = input_files_.insert(file);
+    if (!result.second) // existing element - merge in new information
+    {
+      input_files_.modify(result.first, [&file](InputFile& existing)
+                          {
+                            existing.merge(file);
+                          });
+    }
+
+    return result.first;
   }
 
 
   IdentificationData::ProcessingSoftwareRef
-  IdentificationData::registerDataProcessingSoftware(
-    const DataProcessingSoftware& software)
+  IdentificationData::registerProcessingSoftware(
+    const ProcessingSoftware& software)
   {
-    for (ScoreTypeRef score_ref : software.assigned_scores)
+    if (!no_checks_)
     {
-      if (!isValidReference_(score_ref, score_types_))
+      for (ScoreTypeRef score_ref : software.assigned_scores)
```

Comment:
I guess this is just an iterator? then the copy is fine

Reply:
Yes, all the `...Ref` types are.